ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/SponsoredInteractionCoder.java

```diff
@@ -0,0 +1,41 @@
+package com.mozilla.telemetry.contextualservices;
+
+import org.apache.beam.sdk.coders.*;
+import org.apache.beam.sdk.values.TypeDescriptor;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Map;
+
+public class SponsoredInteractionCoder extends CustomCoder<SponsoredInteraction> {
```

Comment:
I am hoping that we don't need to maintain this coder and can instead use AutoValueSchema like in https://github.com/GoogleCloudPlatform/dataflow-sample-applications/blob/master/retail/retail-java-applications/data-engineering-dept/data-objects/src/main/java/com/google/dataflow/sample/retail/dataobjects/ClickStream.java#L141

See https://beam.apache.org/releases/javadoc/2.19.0/org/apache/beam/sdk/schemas/AutoValueSchema.html

Beam schemas implicitly register a coder.

Reply:
yeah i looked at that and I don't love the custom coder I need a clearer understanding of the difference between schema and coder and how to use the schema properly

Reply:
The Beam programming guide now has some pretty comprehensive docs about schemas. In particular, see https://beam.apache.org/documentation/programming-guide/#inferring-schemas

It's my understanding that when you annotate the class with `@DefaultSchema(AutoValueSchema.class)`, part of the annotation processing is that it generates a row-based coder and adds it to Beam's coder registry. 

I believe the relevant bit where we jump from schema to coder is:

https://github.com/apache/beam/blob/7eeffe05999bb0aea44148e76dedda8482744ea9/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/SchemaCoder.java#L113-L120

So you should never have to explicitly specify a coder for these PCollections.